File: 2018/Thang10-2018/08-10-2018/reduce.pas
Const   ginp='reduce.inp';
        gout='reduce.out';
        maxn=201;
        maxg=40010;

Var     n,maxs,start:longint;
        a:array[0..maxn] of longint;
        b:array[0..2*maxn] of longint;
        d:array[0..maxn,-maxg..maxg] of longint;

Procedure Enter;
Var     i:longint;
Begin
        readln(n,maxs);
        for i:=1 to n do read(a[i]);
End;

Procedure Answer(i,j:longint);
Begin
        if i=0 then exit;
        if d[i,j]=1 then
        begin
                b[i]:=1; j:=j-a[i];
        end else
        begin
                b[i]:=-1; j:=j+a[i];
        end;
        answer(i-1,j);
End;

Procedure Process;
Var     i,j,j1,j2,cou,dcou:longint;
Begin
        d[0,0]:=1;
        for i:=1 to n do
                for j:=-maxg to maxg do
                begin
                        j1:=j+a[i]; j2:=j-a[i];
                        if (j1<=maxg) and (d[i-1,j1]<>0) then d[i,j]:=-1 else
                        if (j2>=-maxg) and (d[i-1,j2]<>0) then d[i,j]:=1 else
                        d[i,j]:=0;
                end;
        answer(n,maxs);
        for i:=n+1 to 2*n do b[i]:=b[i-n];
        for start:=1 to n do
                if (b[start]=1) and (b[start+1]=-1) then break;
        cou:=0; dcou:=0;
        b[start+n]:=b[start+n-1];
        for i:=start+n-1 downto start do
                if b[i]=b[i+1] then inc(cou) else
                begin
                for j:=1 to cou do
                begin
                        if i>n then write(i-n-1,' ')
                        else write(i-dcou-1,' ');
                        if i+j>n then inc(dcou);
                end;
                cou:=1;
                end;
End;

Begin
        Assign(input,ginp); Assign(output,gout);
        Reset(input); Rewrite(output);
        Enter;
        Process;
        Close(input); Close(output);
End.

File: 2019/Thang02_2018/19_02_2019/reduce.cpp
#include <iostream>
#include <stdio.h>
#include <vector>

using namespace std;
//////

const
    int maxn=210, maxg=maxn*maxn;

int n, s;
int a[maxn], d[maxn*2];
bool f[maxn][maxg*2+1];
vector<int> q, ot;
//////

void enter()
{
    scanf("%d%d", &n, &s);
    for (int i=1; i<=n; i++) scanf("%d", &a[i]);
}
//////

void init()
{
    f[0][0+maxg]=1;
    for (int i=1; i<=n; i++)
        for (int j=0; j<=maxg*2; j++)
        {
            if (j-a[i]>=0 && f[i-1][j-a[i]]) f[i][j]=1;
            if (j+a[i]<=maxg*2 && f[i-1][j+a[i]]) f[i][j]=1;
        }
    int i=n, j=s+maxg;
    while (i>0)
    {
        if (j-a[i]>=0 && f[i-1][j-a[i]])
        {
            d[i]=1;
            j=j-a[i];
        } else
        if (j+a[i]<=maxg*2 && f[i-1][j+a[i]])
        {
            d[i]=-1;
            j=j+a[i];
        }
        i--;
    }
    //for (int i=1; i<=n; i++) cout << d[i] << ' ';
}
///////

void process()
{
    ot.clear();
    q.clear();
    d[0]=0;
    for (int i=1; i<=n; i++)
    if (d[i]!=d[i-1]) q.push_back(i);
    if (q.size()==1) return;
    q.push_back(n+1);
    //for (int i=0; i<q.size(); i++) cout << q[i] << ' ';

    for (int i=q.size()-2; i>1; i--)
            for (int j=1; j<=q[i+1]-q[i]; j++) ot.push_back(q[i]-1);
    int dcou=0;
    if (d[1]==1)
    {
        for (int i=1; i<=q[1]-q[0]-1; i++)
        {
            ot.push_back(q[2]-1-dcou);
            dcou++;
        }
        for (int i=1; i<=q[2]-q[1]; i++)
            ot.push_back(1);
    } else
    if (d[1]==-1)
    {
        for (int i=1; i<=q[2]-q[1]-1; i++)
        {
            ot.push_back(q[1]-1);
            dcou++;
        }
        for (int i=1; i<=q[1]-q[0]; i++)
        {
            ot.push_back(q[2]-1-dcou);
            dcou++;
        }
    }
}
///////

int main()
{
    freopen("reduce.inp", "r", stdin);
    freopen("reduce.out", "w", stdout);
    enter();
    init();
    process();
    for (int i=0; i<ot.size(); i++) cout << ot[i]-1 << ' ';
}
